network/p2p/subscription_filter_test.go

```diff
@@ -0,0 +1,119 @@
+package p2p
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/id"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+func TestFilterSubscribe(t *testing.T) {
+	// skip for now due to bug in libp2p gossipsub implementation:
+	// https://github.com/libp2p/go-libp2p-pubsub/issues/449
+	t.Skip()
+
+	identity1, privateKey1 := createID(t, unittest.WithRole(flow.RoleAccess))
+	identity2, privateKey2 := createID(t, unittest.WithRole(flow.RoleAccess))
+	ids := flow.IdentityList{identity1, identity2}
+
+	node1 := createNode(t, identity1.NodeID, privateKey1, createSubscriptionFilterPubsubOption(t, ids))
+	node2 := createNode(t, identity2.NodeID, privateKey2, createSubscriptionFilterPubsubOption(t, ids))
+
+	unstakedKey, err := unittest.NetworkingKey()
+	require.NoError(t, err)
+	unstakedNode := createNode(t, flow.ZeroID, unstakedKey)
+
+	require.NoError(t, node1.AddPeer(context.TODO(), *host.InfoFromHost(node2.Host())))
+	require.NoError(t, node1.AddPeer(context.TODO(), *host.InfoFromHost(unstakedNode.Host())))
+
+	badTopic := engine.TopicFromChannel(engine.SyncCommittee, rootBlockID)
+
+	sub1, err := node1.Subscribe(context.TODO(), badTopic)
+	require.NoError(t, err)
+
+	sub2, err := node2.Subscribe(context.TODO(), badTopic)
+	require.NoError(t, err)
+
+	unstakedSub, err := unstakedNode.Subscribe(context.TODO(), badTopic)
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		return len(node1.pubSub.ListPeers(badTopic.String())) > 0 &&
+			len(node2.pubSub.ListPeers(badTopic.String())) > 0 &&
+			len(unstakedNode.pubSub.ListPeers(badTopic.String())) > 0
+	}, 1*time.Second, 100*time.Millisecond)
+
+	require.Never(t, func() bool {
+		for _, pid := range node1.pubSub.ListPeers(badTopic.String()) {
+			if pid == unstakedNode.Host().ID() {
+				return true
+			}
+		}
+		return false
+	}, 1*time.Second, 100*time.Millisecond)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	testPublish := func(wg *sync.WaitGroup, from *Node, sub *pubsub.Subscription) {
+		data := []byte("hello")
+
+		from.Publish(context.TODO(), badTopic, data)
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		msg, err := sub.Next(ctx)
+		cancel()
+		require.NoError(t, err)
+		require.Equal(t, msg.Data, data)
+
+		ctx, cancel = context.WithTimeout(context.Background(), time.Second)
+		msg, err = unstakedSub.Next(ctx)
+		cancel()
+		require.ErrorIs(t, err, context.DeadlineExceeded)
+
+		wg.Done()
+	}
+
+	// publish a message from node 1 and check that only node2 receives
+	testPublish(&wg, node1, sub2)
+
+	// publish a message from node 2 and check that only node1 receives
+	testPublish(&wg, node2, sub1)
+
+	wg.Wait()
+}
```

Comment:
So, even if libp2p limits the validity of this, here's what happens:
- the subscriptions of the staked nodes to pass,
- the subscription of the unstaked node to fail,
- and, because of the libp2p subscription, the `testPublish` to fail on either of the `Publish` calls at the end.

I surmise we can "split" this test between the subscription logic, otoh, and the `testPublish` test, which should be skipped. This isn't a duplicate of `TestCanSubscribe` because it exercises the unstaked logic.

Reply:
Sure, we can do taht.